crates/bevy_ecs/src/change_detection.rs

```diff
@@ -65,6 +72,24 @@ pub trait DetectChanges {
     /// [`SystemChangeTick`](crate::system::SystemChangeTick)
     /// [`SystemParam`](crate::system::SystemParam).
     fn last_changed(&self) -> u32;
+
+    /// Sets `self` to `value`, if and only if `*self != *value`
+    ///
+    /// `T` is the type stored within the smart pointer (e.g. [`Mut`] or [`ResMut`]).
+    ///
+    /// This is useful to ensure change detection is only triggered when the underlying value
+    /// changes, instead of every time [`DerefMut`] is used.
+    #[inline]
+    fn set_if_neq<T>(&mut self, value: T)
+    where
+        Self: Deref<Target = T> + DerefMut<Target = T>,
+        T: PartialEq,
+    {
+        // This dereference is immutable, so does not trigger change detection
```

Comment:
At first reading, I wasn't a huge fan of the term `immutable` here. However, it seems to be the [standard terminology](https://doc.rust-lang.org/std/ops/trait.Deref.html), so it's fine.